### ISSUE TITLE
Camera pose optimization supported?

### ISSUE THREAD

**Issue by yxie20** [NVlabs/nvdiffrast#5]
Hello all! Thank you again for open-sourcing this powerful tool.

I am wondering that camera pose optimization is trivially supported by nvdiffrast? The rendering pipeline would look like this:
<img src="https://user-images.githubusercontent.com/40676957/99908269-abe8ed00-2caf-11eb-940f-48306f257ac6.png" width="250px">

I went ahead and implemented a training loop, but the training outcome wasn't always successful. Camera either sometimes drifts away or simply doesn't converge. Other times the camera can find the optimal solution. Generally, the camera has a higher chance of drifting away when the initial error is large.

My guess is that anti-aliasing gradients (as opposed to soft rasterization), can only reason at a very local, pixel-level scale and cannot reason at a far distance away from an object. Hence the need for gradient-free greedy optimization in samples/pose.py?

Thank you!

**Comment by s-laine** [comment]
Hi @yxie20 - for geometric optimization (e.g., vertex coordinates or camera position) you should include an antialiasing step at the end of the rendering pipeline. Otherwise you won't obtain gradients related to triangle visibility/coverage, as the coverage is discontinuous w.r.t. vertex positions without antialiasing. The bit about [antialiasing](https://nvlabs.github.io/nvdiffrast/#antialiasing) in the documentation gives more details, and Section 3.6 in the [paper](http://arxiv.org/abs/2011.03277) describes the actual implementation in case you're curious.

**Comment by satyajitghana** [comment]
@yxie20 hey, did it work ?
I've tried pytorch3d for the same, but failed. my loss becomes NaN and the camera drifts away after some time.
I'm thinking of trying out this.
